### ISSUE TITLE
[rush] Add option to allow upper case characters in package names

### ISSUE THREAD

**Issue by james-cordeiro** [microsoft/rushstack#1860]
## Please prefix the issue title with the project name i.e. [rush], [api-extractor] etc.

**Is this a feature or a bug?**

* [x]  Feature (but kind of a bug too)

* [x]  Bug

**Please describe the actual behavior.**

 

https://github.com/microsoft/rushstack/blob/5627415032539c27e93fdddac4edf93d2c87a311/libraries/node-core-library/src/PackageName.ts#L122

 

The above line causes an error when using commands like `rush update` or `rush install` and the npm registry in use is a private one that allows uppercase characters in scope names

 

If the npm registry (from `.npmrc` file) is a private registry where no restriction on this naming convention is required these conditions should not be validated. In fact, I'm pretty sure the rules (including scopes starting with an '@' symbol) in this file are npm registry restrictions/conventions but not necessarily part of the npm spec itself.

 

These restrictions should therefore be removed altogether and allowed to be handled by central registry - it’ll simply error if rules/conventions are not followed during `rush publish` stage. If a project points to the public npm registry then whatever package declared in the package.json file will have to be lowercase otherwise when installing the npm registry it may throw an error (such as 404 status codes) because any package author trying to publish an uppercase scope/name package to the npm registry will fail to be able to even upload in the first place.

 

However, if you declare in your `.npmrc` file a private registry that doesn't enforce these conventions and there is a package on that registry that is private and named with uppercases then rush commands should be able to still download those and the above validation should not throw.

 

Long story short, remove the package name validations as conventions and let the registry itself simply throw if the package can't be retrieved or if trying to publish rush package. This also includes allowing your internal packages (in a rush repo) to be able to declared with uppercase names because in the end it is up to the developer where these packages are going to be `rush publish-ed` to. i.e. if you try to publish to npm registry using `rush publish` with uppercase names then the registry will throw - no need for rush to validate or follow this opinion as it then limits the scope of using rush for private repos in private registries where this rule/convention does not apply

 

Can this, therefore, be removed or is there another sane reason why it needs to be here?

 

Thanks

**Comment by iclanton** [comment]
@octogonz - thoughts on this? It doesn't seem like the Node [module spec](https://nodejs.org/api/modules.html) has an opinion here.

Perhaps we add an option in `rush.json` to allow for a non-strict package name validation mode.

@james-cordeiro - would you be interested in putting together a PR making this change?

**Comment by james-cordeiro** [comment]
Hi @iclanton, I would definitely be interested in putting a PR together. I was planning to in fact but need to find the time to get to it hence, raising the issue if anybody else could pick it up.

If not though and I get the time, I will raise a PR based on the 'add an option' approach within rush.json you mentioned.

Thanks

**Comment by octogonz** [comment]
> Long story short, remove the package name validations as conventions and let the registry itself simply throw if the package can't be retrieved or if trying to publish rush package. 

Sounds like you're advocating to remove ALL validation, not just case.  That can create parsing ambiguities.  For example, TSDoc declaration references use `!` to indicate the end of a package name, so they won't work properly with a package name containing `!` character.  NPM itself uses `@` as a delimiter in many contexts (e.g. `npm install react@3.x`), so allowing `@` in the package name would break that.  

I could go on, but the reason this matters is that parsing ambiguity errors tend to be difficult to understand and a poor experience for end users.  As one example, if your package name is `react@3` and a naive person tries to do `npm install react@3`, it would be very frustrating to get `react` version 3 instead of `react@3`.  Strictness eliminates that sort of problem.

That said, the specific rule regarding uppercase characters is unlikely to cause trouble, because it was actually supported by NPM in ancient times.  So I'm open-minded about that one.

> It doesn't seem like the Node [module spec](https://nodejs.org/api/modules.html) has an opinion here.

The NPM community tends to not publish strict specifications.  For tools the need to provide guarantees of reliable behavior, we we're left to infer an informal spec from "whatever most implementers do." In most cases it works out okay.  The `npmjs.com` registry is a pretty definitive implementer.

> However, if you declare in your `.npmrc` file a private registry that doesn't enforce these conventions and there is a package on that registry that is private and named with uppercases then rush commands should be able to still download those and the above validation should not throw.

I'm super interested in learning more about this scenario.  Why choose to work in a way that is [certain to cause compatibility problems](https://github.com/npm/npm/issues/7351) later down the road?  Why not rename the packages to follow the standard?  There must be a unique explanation, so I'm curious to hear about.  :-)

**Comment by octogonz** [comment]
One other thought: If we do relax the rule, we should not change the default behavior of `PackageName.tryParse()`.  Its purpose is to validate that names are well-formed according to the standard.

But we could consider introducing a "lax" mode that is opt-in, similar to the `validForOldPackages` feature of [validate-npm-package-name](https://github.com/npm/validate-npm-package-name).

**Comment by james-cordeiro** [comment]
Hi @octogonz - completely get your points. Was advocating for relaxing all rules related to registry conventions but in the end for my own use case just need the rule on package names (specifically scopes) to be relaxed, but I get the need for not breaking as well as it also might be an edge case so, I agree with the approach of this being an opt-in setting defined in rush.json. I can work on PR or hoping somebody else can get to this.

In mean time,

>I'm super interested in learning more about this >scenario. Why choose to work in a way that is >certain to cause compatibility problems later down >the road? Why not rename the packages to follow >the standard? There must be a unique explanation, >so I'm curious to hear the background. :-)

....background here is in normal cirmstances, I would simply rename the packages to be lower case - agreed.

Unfortunately though, the private registry I'm working with includes a dozen teams (globally) throughout my organisation publishing and maintaining 100s of internal packages. Then there are additional (many) internal product teams consuming these packages and to ask both the app developers and package authors to rename everything (all scopes are in CAPS throughout org to represent each team or group) to be lower case will be impractical at this time. One of my goals is to get the organisation to start moving towards a single large monorepo but in the mean time as a member of one of those app teams (with multiple package modules of our own to publish) I'd like to first POC rush as a tool to move forward with. However, as packages we already consume are named with CAPS for scopes; asking the package authoring teams to change the package names mean coordinating other internal consumers to follow the new convention if they still want to receive updates/upgrades for the future. This should be ok also (change packages names - to your point around future compatibility issues) but it would also take time given all manners of approval processes and co-ordination between many multiple teams - major versions move fast in our case so again, impractical to go down this route as yes, we have now have the compatibility issue "down the road" already as you mentioned.

Now, I get this will be a very specific use case and might even be a single use case but I would still challenge you that it's still a Rush "thing" to solve and not a consumer "thing" (by that, just change the name) as in the end, the compatibility issue you mention is only related to the npm registry but not all registries follow this rule (I believe not even private registry within Azure DevOps follows this either does it? - we use Nexus btw) and in fact, by "binding" to this convention (albeit it's so common it could be a standard) it actually hinders (like in my case) teams to use a tool like rush within private registries owned and maintained by large organisations.

One final note, there is no "technical problem" from changing our org package names but I guess it's one of those "somebody wanted it this way" so it stuck issues. However, as opensource custodians, being open for all, I would personally like to see Rush not impose that this rule HAS to be followed, but as it's so widely adopted it makes sense to be "on" be default, so long as there is an escape hatch - that being the relax setting as suggested. 

Thank you for your comments, really well informed. :)

**Comment by octogonz** [comment]
> Unfortunately though, the private registry I'm working with includes a dozen teams (globally) throughout my organisation publishing and maintaining 100s of internal packages. Then there are additional (many) internal product teams consuming these packages and to ask both the app developers and package authors to rename everything (all scopes are in CAPS throughout org to represent each team or group) to be lower case will be impractical at this time.

Seems like a pretty good reason.  And your group sounds like the exact audience that Rush was built for. 👍

I'll make the PR myself.

**Comment by octogonz** [comment]
For some trivia, here's a list of all non-conforming package names that were published to the NPM registry before they implemented stricter name validation (queried from [all-the-package-names](https://www.npmjs.com/package/all-the-package-names)):

```
    tinytim
 optimist
?
@(._.)/assign
@(._.)/execute
@(._.)/oooooo
@____/_
@agm-universal/_dev
@angular-toolkit/_utils
@billti/_foo
@bre!zh/emitter
@chappa'ai/get-next-rc
@chappa'ai/next-package-version
@chappa'ai/update-packages
@es-git/__template__
@ganuz/_get-metadata-map
@ganuz/_map
@ganuz/_weak-map
@jskit/_demo
@ng-toolkit/_utils
@npmcorp/.github
@promises/_create-checks-boolean
@promises/_create-filter
@promises/_create-for-each-series
@promises/_create-map
@promises/_create-reduce
@promises/_curry
@promises/_keys
@ramda/__
@rcp/_types
@romeria/_
@s9a/...npm
@sz.frankyy/__ajax
@zemnmez/.
@~962464/hello-wasm
@~962464/test
@~alexwang/gg-editor-alex
@~banchen/yj-img-uploader
@~graphite/segmenter
@~graphite/segmenter-objectid
@~gregoryalary/n-map
@~honchy/node-util
@~lisfan/event-queues
@~lisfan/gulp-inject-string
@~lisfan/logger
@~lisfan/storage
@~lisfan/timer
@~lisfan/validation
@~lisfan/vue-image-loader
@~lisfan/vue-image-placeholder
@~lisfan/vue-logger
@~lisfan/vue-upyun-image-format
@~sunsimiao/cj-dnd
@~sunsimiao/cj-dnd-backend
@~sunsimiao/cj-dnd-core
@~sunsimiao/cj-react-dnd
@~sunsimiao/html-to-draftjs
@~sunsimiao/module-1
@~sunsimiao/module-2
@~xl/generator-vue-template
JQuery UI
__namespace__class-plugin-should
abcd" &gt;onmouseover=alert(1)\
babel-preset-stage-0 
bluebird 
brain***_games***
dcl-bootstrap-attributes 
dustjs-linkedin 
eslint 
expect.js 
extract-text-webpack-plugin 
f*
foo!
foo~
grunt-browserify 
grunt-coffeelint 
grunt-contrib-copy 
grunt-contrib-cssmin 
grunt-contrib-uglify 
grunt-contrib-yuidoc 
grunt-mocha-test:
gulp-handlebars 
hahahaha||whoami||
handlebars 
hash!
hbsfy 
highcharts-*
html-to-vdom 
iana-http-methods 
is-array"asd
jasmine-expect 
jsonfile 
koa@2
level 
nfix 
ng-html2js 
node-sass 
nodeunit 
pybee!batavia
react-lensed-state 
run-sequence 
sass 
uuid 
xmlhttprequest 
yargs 
{}
```

I'm quite surprised at (a) how few there are from a data set of >1 million names, (b) how malformed they are, and (c) that only one of them actually contains upper-case characters.

**Comment by octogonz** [comment]
Proposed `rush.json` setting:

```js
  /**
   * Today the npmjs.com registry enforces fairly strict naming rules for packages, but in the early
   * days there was no standard and hardly any enforcement.  A few large legacy projects are still using
   * nonstandard package names, and private registries sometimes allow it.  Set "allowMostlyStandardPackageNames"
   * to true to relax Rush's enforcement of package names.  This allows upper case letters and in the future may
   * relax other rules, however we want to minimize these exceptions.  Many popular tools use certain punctuation
   * characters as delimiters, based on the assumption that they will never appear in a package name; thus if we relax
   * the rules too much it is likely to cause very confusing malfunctions.
   *
   * The default value is false.
   */
  "allowMostlyStandardPackageNames": false,
```

**Comment by james-cordeiro** [comment]
> For some trivia, here's a list of all non-conforming package names that were published to the NPM registry before they implemented stricter name validation (queried from [all-the-package-names](https://www.npmjs.com/package/all-the-package-names)):
> 
> ```
> "    tinytim"
> " optimist"
> "?"
> "@(._.)/assign"
> "@(._.)/execute"
> "@(._.)/oooooo"
> "@____/_"
> "@agm-universal/_dev"
> "@angular-toolkit/_utils"
> "@billti/_foo"
> "@bre!zh/emitter"
> "@chappa'ai/get-next-rc"
> "@chappa'ai/next-package-version"
> "@chappa'ai/update-packages"
> "@es-git/__template__"
> "@ganuz/_get-metadata-map"
> "@ganuz/_map"
> "@ganuz/_weak-map"
> "@jskit/_demo"
> "@ng-toolkit/_utils"
> "@npmcorp/.github"
> "@promises/_create-checks-boolean"
> "@promises/_create-filter"
> "@promises/_create-for-each-series"
> "@promises/_create-map"
> "@promises/_create-reduce"
> "@promises/_curry"
> "@promises/_keys"
> "@ramda/__"
> "@rcp/_types"
> "@romeria/_"
> "@s9a/...npm"
> "@sz.frankyy/__ajax"
> "@zemnmez/."
> "@~962464/hello-wasm"
> "@~962464/test"
> "@~alexwang/gg-editor-alex"
> "@~banchen/yj-img-uploader"
> "@~graphite/segmenter"
> "@~graphite/segmenter-objectid"
> "@~gregoryalary/n-map"
> "@~honchy/node-util"
> "@~lisfan/event-queues"
> "@~lisfan/gulp-inject-string"
> "@~lisfan/logger"
> "@~lisfan/storage"
> "@~lisfan/timer"
> "@~lisfan/validation"
> "@~lisfan/vue-image-loader"
> "@~lisfan/vue-image-placeholder"
> "@~lisfan/vue-logger"
> "@~lisfan/vue-upyun-image-format"
> "@~sunsimiao/cj-dnd"
> "@~sunsimiao/cj-dnd-backend"
> "@~sunsimiao/cj-dnd-core"
> "@~sunsimiao/cj-react-dnd"
> "@~sunsimiao/html-to-draftjs"
> "@~sunsimiao/module-1"
> "@~sunsimiao/module-2"
> "@~xl/generator-vue-template"
> "JQuery UI"
> "__namespace__class-plugin-should"
> "abcd\" &gt;onmouseover=alert(1)\\"
> "babel-preset-stage-0 "
> "bluebird "
> "brain***_games***"
> "dcl-bootstrap-attributes "
> "dustjs-linkedin "
> "eslint "
> "expect.js "
> "extract-text-webpack-plugin "
> "f*"
> "foo!"
> "foo~"
> "grunt-browserify "
> "grunt-coffeelint "
> "grunt-contrib-copy "
> "grunt-contrib-cssmin "
> "grunt-contrib-uglify "
> "grunt-contrib-yuidoc "
> "grunt-mocha-test:"
> "gulp-handlebars "
> "hahahaha||whoami||"
> "handlebars "
> "hash!"
> "hbsfy "
> "highcharts-*"
> "html-to-vdom "
> "iana-http-methods "
> "is-array\"asd"
> "jasmine-expect "
> "jsonfile "
> "koa@2"
> "level "
> "nfix "
> "ng-html2js "
> "node-sass "
> "nodeunit "
> "pybee!batavia"
> "react-lensed-state "
> "run-sequence "
> "sass "
> "uuid "
> "xmlhttprequest "
> "yargs "
> "{}"
> ```
> 
> I'm quite surprised at (a) how few there are from a data set of >1 million names, (b) how malformed they are, and (c) that only one of them actually contains upper-case characters.

Love this! ....and only one has an uppercase! ....man, that must have sucked JQuery's "branding" angle too!

**Comment by octogonz** [comment]
Here is a PR: https://github.com/microsoft/rushstack/pull/1879

**Comment by octogonz** [comment]
@james-cordeiro  This was released with Rush 5.23.4

**Comment by james-cordeiro** [comment]
@octogonz hey - thanks. Sorry, been off radar. Will give it a go asap.

**Comment by Zemelia** [comment]
Hello, but does it possible to use this for an api-extractor?
As I didn't find a way...
